File: android/src/main/java/expo/modules/superwallexpo/json/PaywallSkippedReason.kt
package expo.modules.superwallexpo.json

import com.superwall.sdk.paywall.presentation.internal.state.PaywallSkippedReason

  fun PaywallSkippedReason.toJson(): Map<String, Any> {
      val map = mutableMapOf<String, Any>()

      when (this) {
        is PaywallSkippedReason.Holdout -> {
          map["type"] = "Holdout"
          map["experiment"] = Experiment.toJson(experiment)
        }

        is PaywallSkippedReason.NoAudienceMatch -> {
          map["type"] = "NoAudienceMatch"
        }

        is PaywallSkippedReason.PlacementNotFound -> {
          map["type"] = "PlacementNotFound"
        }

        is PaywallSkippedReason.UserIsSubscribed -> {
          map["type"] = "UserIsSubscribed"
        }
      }

  return map
}

File: src/compat/lib/PaywallSkippedReason.ts
import { Experiment } from "./Experiment"

/**
 * @category Models
 * @since 0.0.15
 * Abstract class representing the reason why a paywall was skipped.
 */
export abstract class PaywallSkippedReason extends Error {
  constructor(message?: string) {
    super(message)
    this.name = new.target.name
  }

  static fromJson(json: any): PaywallSkippedReason {
    switch (json.type) {
      case "Holdout": {
        const experiment = Experiment.fromJson(json.experiment)
        return new PaywallSkippedReasonHoldout(experiment)
      }
      case "NoAudienceMatch":
        return new PaywallSkippedReasonNoAudienceMatch()
      case "PlacementNotFound":
        return new PaywallSkippedReasonPlacementNotFound()
      case "UserIsSubscribed":
        return new PaywallSkippedReasonUserIsSubscribed()
      default:
        throw new Error("Unknown PaywallSkippedReason type")
    }
  }
}

/**
 * @category Models
 * @since 0.0.15
 * Paywall skipped due to user being in a holdout group.
 */
export class PaywallSkippedReasonHoldout extends PaywallSkippedReason {
  experiment: Experiment

  constructor(experiment: Experiment) {
    super("Holdout")
    this.experiment = experiment
  }
}

/**
 * @category Models
 * @since 0.0.15
 * Paywall skipped due to no audience match.
 */
export class PaywallSkippedReasonNoAudienceMatch extends PaywallSkippedReason {
  constructor() {
    super("NoAudienceMatch")
  }
}

/**
 * @category Models
 * @since 0.0.15
 * Paywall skipped due to placement not being found.
 */
export class PaywallSkippedReasonPlacementNotFound extends PaywallSkippedReason {
  constructor() {
    super("PlacementNotFound")
  }
}

/**
 * @category Models
 * @since 0.0.15
 * Paywall skipped due to user already being subscribed.
 */
export class PaywallSkippedReasonUserIsSubscribed extends PaywallSkippedReason {
  constructor() {
    super("UserIsSubscribed")
  }
}
